### PR TITLE
Added ElementSetTool base class.

### PR DESCRIPTION
**Pull request by bbastings**
This base class is useful for any tool that wants to operation on sets of elements/support selection sets. Next step is to update modify tool examples/create some basic modification tools that use EditCommand instead of IModelWriteRpcInterface/ElementEditor3d.

### CODE REVIEW

File: core/frontend/src/tools/ElementSetTool.ts [hunk 1]
```diff
@@ -6,141 +6,51 @@
...
 export interface GroupMark {
   start: number;
   source: ModifyElementSource;
 }
 
-/** @internal */
+/** @alpha */
 export class ElementAgenda {
-  public readonly elements: string[] = [];
+  public readonly elements: Id64Array = [];
```
Comment (pmconne) [inline core/frontend/src/tools/ElementSetTool.ts:48]: Member functions like `add` and `remove` do additional stuff with e.g. hilite and group marks. Is it cool if people can just modify this array directly?

File: core/frontend/src/tools/ElementSetTool.ts [hunk 2]
```diff
@@ -307,25 +204,705 @@ export class ElementAgenda {
...
+          this.onReinitialize();
+          return EventHandled.Yes;
+        }
+
+        this.setupAndPromptForNextAction();
+        return EventHandled.No;
+      }
+    }
+
+    IModelApp.accuSnap.resetButton(); // eslint-disable-line @typescript-eslint/no-floating-promises
```
Comment (pmconne) [inline core/frontend/src/tools/ElementSetTool.ts:667]: Is there a reason not to `await` this?

File: core/frontend/src/tools/ElementSetTool.ts [hunk 3]
```diff
@@ -150,44 +60,41 @@ export class ElementAgenda {
...
-  /** Mark all entries in this agenda as being hilited. */
-  public hilite() {
-    if (HilitedState.Yes === this.hilitedState)
-      return;
-
-    this.setEntriesHiliteState(this.hiliteOnAdd); // make sure all entries have their hilite flag on.
-    this.hilitedState = HilitedState.Yes;
-  }
-
+  /** Return true if elementId is already in this ElementAgenda. */
```
Comment (pmconne) [inline core/frontend/src/tools/ElementSetTool.ts:97]: `add` calls this in a loop to make sure a given element only appears once in the array. Doesn't the agenda really want to be a `Set` not an array?

File: core/frontend/src/tools/ElementSetTool.ts (hunk 2 above)
Comment (bbastings) [reply core/frontend/src/tools/ElementSetTool.ts:667]: Just being consistent with SelectTool, there's no need to await...

File: core/frontend/src/tools/ElementSetTool.ts (hunk 3 above)
Comment (bbastings) [reply core/frontend/src/tools/ElementSetTool.ts:97]: No, the array is important for order, ex. tool wants 1st to be a solid, 2nd to be a surface, etc.

Comment (pmconne) [reply core/frontend/src/tools/ElementSetTool.ts:97]: `Set` preserves order.

File: core/frontend/src/tools/ElementSetTool.ts [hunk 4]
```diff
@@ -6,141 +6,51 @@
...
-/** @internal */
+/** @alpha */
 export interface GroupMark {
   start: number;
   source: ModifyElementSource;
 }
 
-/** @internal */
+/** @alpha */
 export class ElementAgenda {
```
[the remark this answers is not in the document]
Comment (bbastings) [reply core/frontend/src/tools/ElementSetTool.ts:47]: Consistency/convenience w/SelectionSet...

[the remark this answers is not in the document]
Comment (pmconne) [reply core/frontend/src/tools/ElementSetTool.ts:47]: I really weary of that reasoning - "we made uninformed decisions years ago, so we must propagate them now."

File: core/frontend/src/tools/ElementSetTool.ts (hunk 1 above)
Comment (bbastings) [reply core/frontend/src/tools/ElementSetTool.ts:48]: I'll add a comment that it's better not to change directly. The "get candidate" methods on ElementSetTool should steer people towards the right way to modify the agenda.

File: core/frontend/src/tools/ElementSetTool.ts (hunk 3 above)
Comment (bbastings) [reply core/frontend/src/tools/ElementSetTool.ts:103]: Only being able to iterate in insert order seemed inconvenient, and I'd have to rework GroupMark?

**Review by pmconne — APPROVED** [review]

**Pull request merged**